### PR TITLE
Add /scale subresource to CRD and replicas to various parts of CRD.

### LINKED ISSUE #1592: Add Scale SubResource to CRD
This would allow `kubectl scale <name>` to be used with our CRD.

### LINKED ISSUE #1598: Be able to scale SeldonDeployment objects using kubectl
kubectl allows one to scale their normal Deployment objects using the 'kubectl scale' command by setting the '--replicaset' command line argument to the number of Pods you want the underlying ReplicaSet to achieve.

However, with SeldonDeployment objects, you can't do this. In fact, in 1.0+, you can't even use `kubectl scale` on the underlying Deployment object that backs the SeldonDeployment one to up the number of prediction server Pods.

What you have to do is delete and redeploy the SeldonDeployment object and up the number of 'replicas' in the original YAML/JSON deployment file.

I think it would be best if SeldonDeployment objects could scale like regular Deployment ones as their configuration and management idioms should be fairly the same.

(I also think both 'seldonDeployment' and 'seldondeploy' should work with kubectl too, just like deployment).

### PR DESCRIPTION
**Pull request by cliveseldon**
Fixes #1592 
Fixes #1598 

 * Adds /scale subresource to CRD
 * Adds `defaultReplicas` to spec
   * Any predictor replicas not set will be set to spec.defaultReplicas
 * Add `.spec.predictors[].componentSpecs[].replicas`
    * Allows replicas to be set just for the deployment created from the componentSpec
 * Adds `.spec.predictors[].svcOrchSpec.replicas`
 * Changes predictor.replicas to pointer *int32
 * Add defaultReplicas to status
 * Updates Autoscaling example and adds as notebook test
 * Add new scale notebook to test replica settings and scaling via kubectl
 * Adds docs for above
 
Will allow for SeldonDeployment `seldon-model` for example,

```
kubectl scale --replicas=2 sdep/seldon-model
```

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **cliveseldon**

If they are not already assigned, you can assign the PR to them by writing `/assign @cliveseldon` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Sun Mar 29 18:40:03 UTC 2020
The logs for [lint] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=2

**Comment by seldondev** [comment]
Sun Mar 29 18:40:06 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=1

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | 4c963a4bf808578c61a6332db304b37fb047d009 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Sun Mar 29 18:42:13 UTC 2020
The logs for [pr-build] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=3

**Comment by seldondev** [comment]
Sun Mar 29 18:42:13 UTC 2020
The logs for [lint] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=4

**Comment by seldondev** [comment]
Sun Mar 29 18:42:23 UTC 2020
The logs for [integration] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=5

**Comment by seldondev** [comment]
Sun Mar 29 21:41:50 UTC 2020
The logs for [notebooks] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=6

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | cf0744a0a878a82e605dfec61a20e3dc26a226d7 | [link]() | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Mon Mar 30 15:00:22 UTC 2020
The logs for [lint] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=8

**Comment by seldondev** [comment]
Mon Mar 30 15:00:39 UTC 2020
The logs for [pr-build] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=7

**Comment by seldondev** [comment]
Mon Mar 30 15:02:23 UTC 2020
The logs for [notebooks] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=9

**Comment by seldondev** [comment]
Mon Mar 30 15:49:12 UTC 2020
The logs for [integration] [10] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/10.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=10

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 0b3d4ef5236721e2568fae6d253a79785f4ee480 | [link]() | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Mon Mar 30 16:30:19 UTC 2020
The logs for [lint] [12] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/12.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=12

**Comment by seldondev** [comment]
Mon Mar 30 16:30:21 UTC 2020
The logs for [pr-build] [11] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/11.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=11

**Comment by seldondev** [comment]
Mon Mar 30 16:30:51 UTC 2020
The logs for [notebooks] [13] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/13.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=13

**Comment by seldondev** [comment]
Mon Mar 30 16:32:33 UTC 2020
The logs for [integration] [14] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/14.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=14

**Comment by seldondev** [comment]
Mon Mar 30 17:00:45 UTC 2020
The logs for [pr-build] [15] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/15.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=15

**Comment by seldondev** [comment]
Mon Mar 30 17:00:59 UTC 2020
The logs for [integration] [18] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/18.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=18

**Comment by seldondev** [comment]
Mon Mar 30 17:01:04 UTC 2020
The logs for [lint] [16] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/16.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=16

**Comment by seldondev** [comment]
Mon Mar 30 17:01:03 UTC 2020
The logs for [notebooks] [17] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/17.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=17

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 0b3d4ef5236721e2568fae6d253a79785f4ee480 | [link]() | `/test notebooks`
integration | 650ed506ea97cb0bbca4fe97b5bdc1cc13fbac38 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 650ed506ea97cb0bbca4fe97b5bdc1cc13fbac38 | [link]() | `/test integration`
notebooks | 5ed37bf638bce9f9c4baf2dc8326f594564d6705 | [link]() | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 650ed506ea97cb0bbca4fe97b5bdc1cc13fbac38 | [link]() | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: testing/scripts/metrics.yaml [hunk 1]
```diff
@@ -0,0 +1,138 @@
+kind: ClusterRole
```
Comment (adriangonz) [inline testing/scripts/metrics.yaml:1]: nit: should we move `metrics.yaml` to the `./testing/resources/` folder?

File: operator/apis/machinelearning/v1/seldondeployment_types.go [hunk 2]
```diff
@@ -394,18 +397,19 @@ type SeldonDeploymentStatus struct {
...
 // +genclient:noStatus
 // +kubebuilder:object:root=true
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 // +kubebuilder:storageversion
-
 // SeldonDeployment is the Schema for the seldondeployments API
 // +k8s:openapi-gen=true
 // +kubebuilder:resource:shortName=sdep
 // +kubebuilder:subresource:status
+// +kubebuilder:subresource:scale:specpath=.spec.defaultReplicas,statuspath=.status.defaultReplicas
```
Comment (adriangonz) [inline operator/apis/machinelearning/v1/seldondeployment_types.go:412]: It's crazy that this is all you need to add support for `scale`!

File: doc/source/index.rst [hunk 3]
```diff
@@ -95,7 +95,7 @@ Documentation Index
    Custom Metrics <analytics/custom_metrics.md>
    Payload Logging with ELK <analytics/logging.md>
    Distributed Tracing with Jaeger <graph/distributed-tracing.md>
-   Autoscaling in Kubernetes <graph/autoscaling.md>
```
Comment (adriangonz) [inline doc/source/index.rst:98]: This is not something critical, just a thought. 

At some point in the future, would it make sense to consider redirects whenever we remove / rename paths? For example, in this case:

- https://docs.seldon.io/projects/seldon-core/en/latest/graph/autoscaling.html => https://docs.seldon.io/projects/seldon-core/en/latest/graph/scaling.html

Is this something supported by `readthedocs`?

File: operator/apis/machinelearning/v1/seldondeployment_types.go [hunk 4]
```diff
@@ -196,20 +196,21 @@ func GetContainerServiceName(mlDepName string, predictorSpec PredictorSpec, c *v
...
-	Protocol    Protocol          `json:"protocol,omitempty" protobuf:"bytes,6,opt,name=protocol"`
-	Transport   Transport         `json:"transport,omitempty" protobuf:"bytes,7,opt,name=transport"`
+	Name            string            `json:"name,omitempty" protobuf:"string,1,opt,name=name"`
+	Predictors      []PredictorSpec   `json:"predictors" protobuf:"bytes,2,opt,name=name"`
+	OauthKey        string            `json:"oauth_key,omitempty" protobuf:"string,3,opt,name=oauth_key"`
+	OauthSecret     string            `json:"oauth_secret,omitempty" protobuf:"string,4,opt,name=oauth_secret"`
+	Annotations     map[string]string `json:"annotations,omitempty" protobuf:"bytes,5,opt,name=annotations"`
+	Protocol        Protocol          `json:"protocol,omitempty" protobuf:"bytes,6,opt,name=protocol"`
+	Transport       Transport         `json:"transport,omitempty" protobuf:"bytes,7,opt,name=transport"`
+	DefaultReplicas *int32            `json:"defaultReplicas,omitempty" protobuf:"bytes,8,opt,name=defaultReplicas"`
```
[the remark this answers is not in the document]
Comment (cliveseldon) [reply operator/apis/machinelearning/v1/seldondeployment_types.go]: yes make sense

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 650ed506ea97cb0bbca4fe97b5bdc1cc13fbac38 | [link]() | `/test notebooks`
lint | 147ec4ab6114886c6678a10c1f771ae52b2028d5 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 650ed506ea97cb0bbca4fe97b5bdc1cc13fbac38 | [link]() | `/test notebooks`
lint | 147ec4ab6114886c6678a10c1f771ae52b2028d5 | [link]() | `/test lint`
pr-build | 147ec4ab6114886c6678a10c1f771ae52b2028d5 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: testing/scripts/metrics.yaml (hunk 1 above)
Comment (cliveseldon) [reply testing/scripts/metrics.yaml:1]: Not sure as we still have istio-1.4.2.yaml in this folder. These two yamls are part of the cluster setup.

File: operator/apis/machinelearning/v1/seldondeployment_types.go (hunk 2 above)
Comment (cliveseldon) [reply operator/apis/machinelearning/v1/seldondeployment_types.go]: CRD power!

File: doc/source/index.rst (hunk 3 above)
Comment (cliveseldon) [reply doc/source/index.rst:98]: we can add to https://readthedocs.org/dashboard/seldon-core/redirects/ when published

**Comment by seldondev** [comment]
Mon Apr  6 08:25:38 UTC 2020
The logs for [lint] [22] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/22.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=22

File: doc/source/index.rst (hunk 3 above)
Comment (adriangonz) [reply doc/source/index.rst:97]: That's great to know! I didn't know we had that available.

**Comment by seldondev** [comment]
Mon Apr  6 09:13:32 UTC 2020
The logs for [integration] [23] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1633/23.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1633 --build=23

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/SeldonIO/seldon-core/pull/1633#issuecomment-609784118" title="Approved">adriangonz</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [adriangonz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**